packages/python/accern_data/accern_data.py

```diff
@@ -273,18 +277,49 @@ def do_init(self, indicator: ProgressIndicator) -> None:
         fname = self.get_path(self._is_by_day)
         indicator.log(f"current file is {fname}")
 
+    def _write_cols(self) -> None:
+        fname = self.get_path(is_by_day=self._is_by_day)
+        if self._cols:
+            with open(f"{fname}~.columns", "w") as fout:
+                fout.write(",".join(self._cols))
+
+    def get_tmp_file_name(self, fname: str) -> str:
+        return f"{fname}.~tmp"
+
     def add_result(self, signal: pd.DataFrame) -> None:
         fname = self.get_path(is_by_day=self._is_by_day)
+        tmp_fname = self.get_tmp_file_name(fname)
         if self._cols is None:
-            signal.to_csv(fname, index=False, header=True, mode="w")
-            self._cols = signal.columns
+            self._cols = signal.columns.to_list()
+            signal.to_csv(tmp_fname, index=False, header=True, mode="w")
         else:
+            if len(signal.columns) <= len(self._cols):
+                for col in set(self._cols).difference(signal.columns):
+                    signal[col] = None
+            else:
+                new_cols = set(signal.columns).difference(self._cols)
+                self._cols += new_cols
             signal[self._cols].to_csv(
-                fname, index=False, header=False, mode="a")
+                tmp_fname, index=False, header=False, mode="a")
 
-    def finish_day(self, indicator: ProgressIndicator) -> None:
+    def finish_day(
+            self,
+            indicator: ProgressIndicator,
+            force_finish: bool = False) -> None:
         # csv files are saved by add_result
-        if self._is_by_day:
+        fname = self.get_path(is_by_day=self._is_by_day)
+        tmp_fname = self.get_tmp_file_name(fname)
+
+        if (self._is_by_day or force_finish) and self._cols:
+            with open(fname, "w") as file, open(tmp_fname, "r") as tmp_csv:
+                csv_reader = csv.reader(tmp_csv)
+                csv_writer = csv.writer(file)
+                next(csv_reader)  # skip header
+                csv_writer.writerow(self._cols)
+                for row in csv_reader:
+                    csv_writer.writerow(
+                        row + [None] * (len(self._cur_date) - len(row)))
```

Comment:
create tests where there are fewer columns at the beginning then more columns then fewer columns in the middle and then more columns again